src/core/server/legacy/config/ensure_valid_configuration.ts

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { getUnusedConfigKeys } from '../../../../legacy/server/config/get_unused_config_keys';
+import { ConfigService } from '../../config';
+import { LegacyServiceDiscoverPlugins } from '../legacy_service';
+
+export async function ensureValidConfiguration(
+  configService: ConfigService,
+  { pluginSpecs, disabledPluginSpecs, pluginExtendedConfig, settings }: LegacyServiceDiscoverPlugins
+) {
+  const unusedConfigKeys = await getUnusedConfigKeys(
+    await configService.getUsedPaths(),
+    pluginSpecs,
+    disabledPluginSpecs,
+    settings,
+    pluginExtendedConfig
+  );
+
+  if (unusedConfigKeys.length > 0) {
+    const message = `Unknown configuration key(s): ${unusedConfigKeys
+      .map(key => `"${key}"`)
+      .join(', ')}. Check for spelling errors and ensure that expected plugins are installed.`;
+    throw new InvalidConfigurationError(message);
+  }
+}
+
+class InvalidConfigurationError extends Error {
+  public cause?: Error;
+  public code = 'InvalidConfig';
+  public processExitCode = 64;
+
+  constructor(message: string) {
+    super(message);
```

Comment:
Don't we have to add the Object prototype extension as in https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work ?

Reply:
I saw that in another Error extension, but I wasn't sure exactly what that was for. I guess I should do it then?

Reply:
Limitations are:
- methods may be undefined on objects returned by constructing these subclasses, so calling sayHello will result in an error.
- instanceof will be broken between instances of the subclass and their instances, so (new FooError()) instanceof FooError will return false.

Doesn't seem like impacting our usage of this type. Still better to do the workaround than not I guess. Thanks!

Reply:
Yes, I saw it is not really impacting our current usage. I'm more concerned about the fact that `instanceof` is broken if that line is not added. But we are not exporting it either, so I guess if we ever want to catch it, we'll need to change this file anyway :)

Reply:
Added the line just in case of, seems like the good practice. Still hope this will be fixed in TS a some point though, this seems like a workaround.